kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,624 @@
+            Error::InternalError(format!(
+                "Wrong number of AddRemoveDedupVisitor getters: {}",
+                getters.len()
+            ))
+        );
+
+        for i in 0..row_count {
+            if !self.selection_vector[i] {
+                continue;
+            }
+            if getters[0].get_str(i, "add.path")?.is_some() {
+                self.selection_vector[i] = !self.has_cdc_action;
+            } else if let Some(path) = getters[1].get_str(i, "remove.path")? {
+                self.selection_vector[i] =
+                    !self.has_cdc_action && !self.remove_dvs.contains_key(path)
+            } else {
+                self.selection_vector[i] = getters[2].get_str(i, "cdc.path")?.is_some()
+            };
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use crate::scan::state::DvInfo;
+    use crate::schema::{DataType, StructField, StructType};
+    use itertools::Itertools;
+    use object_store::local::LocalFileSystem;
+    use object_store::ObjectStore;
+    use serde::Serialize;
+    use std::collections::{HashMap, HashSet};
+    use std::path::Path;
+    use std::sync::Arc;
+    use tempfile::TempDir;
+    use tokio::runtime::Runtime;
+
+    use super::{get_add_transform_expr, LogReplayScanner};
+    use crate::actions::{get_log_add_schema, Add, Cdc, CommitInfo, Metadata, Protocol, Remove};
+    use crate::engine::sync::SyncEngine;
+    use crate::log_segment::LogSegment;
+    use crate::path::ParsedLogPath;
+    use crate::scan::log_replay::SCAN_ROW_DATATYPE;
+    use crate::{DeltaResult, Engine, Version};
+    use test_utils::delta_path_for_version;
+
+    #[derive(Serialize)]
+    #[allow(unused)]
+    enum Action {
+        #[serde(rename = "add")]
+        Add(Add),
+        #[serde(rename = "remove")]
+        Remove(Remove),
+        #[serde(rename = "cdc")]
+        Cdc(Cdc),
+        #[serde(rename = "metaData")]
+        Metadata(Metadata),
+        #[serde(rename = "protocol")]
+        Protocol(Protocol),
+        #[serde(rename = "commitInfo")]
+        CommitInfo(CommitInfo),
+    }
+
+    impl From<Add> for Action {
+        fn from(value: Add) -> Self {
+            Self::Add(value)
+        }
+    }
+    impl From<Remove> for Action {
+        fn from(value: Remove) -> Self {
+            Self::Remove(value)
+        }
+    }
+    impl From<Cdc> for Action {
+        fn from(value: Cdc) -> Self {
+            Self::Cdc(value)
+        }
+    }
+    impl From<Metadata> for Action {
+        fn from(value: Metadata) -> Self {
+            Self::Metadata(value)
+        }
+    }
+    impl From<Protocol> for Action {
+        fn from(value: Protocol) -> Self {
+            Self::Protocol(value)
+        }
+    }
+    impl From<CommitInfo> for Action {
+        fn from(value: CommitInfo) -> Self {
+            Self::CommitInfo(value)
+        }
+    }
+
+    struct MockTable {
+        commit_num: u64,
+        store: Arc<LocalFileSystem>,
+        runtime: Runtime,
```

Comment:
I think you can just make the functions you want async async and then use `#[tokio::test]` on any tests that want to use a `MockTable`